File: typescript/src/templateEngine/Blocks/codeBlock.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { ILogger } from '../../utils/logger';
import { ContextVariables, ISKFunction, SKContext, SKFunctionExtensions } from '../../orchestration';
import { IFunctionRegistryReader } from '../../registry';
import { Block, BlockTypes } from './block';
import { VarBlock } from './varBlock';

/**
 * @private
 */
export class CodeBlock extends Block {
    private _validated: boolean = false;

    constructor(public content: string, log: ILogger) {
        super(log);
    }

    public get type(): BlockTypes {
        return BlockTypes.Code;
    }

    public isValid(): { valid: boolean; error: string } {
        let valid = true;
        let error = '';

        const partsToValidate = this.content.split(/[ \t\r\n]+/).filter((x) => x.trim() !== '');

        for (let index = 0; index < partsToValidate.length; index++) {
            const part = partsToValidate[index];

            if (index === 0) {
                // There is only a function name
                if (VarBlock.hasVarPrefix(part)) {
                    error = `Variables cannot be used as function names [\`${part}\`]`;
                    this.log.error(error);
                    valid = false;
                }

                if (!/^[a-zA-Z0-9_.]*$/.test(part)) {
                    error = `The function name \`${part}\` contains invalid characters`;
                    this.log.error(error);
                    valid = false;
                }
            } else {
                // The function has parameters
                if (!VarBlock.hasVarPrefix(part)) {
                    error = `\`${part}\` is not a valid function parameter: parameters must be variables.`;
                    this.log.error(error);
                    valid = false;
                }

                if (part.length < 2) {
                    error = `\`${part}\` is not a valid variable.`;
                    this.log.error(error);
                    valid = false;
                }

                if (!VarBlock.isValidVarName(part.substring(1))) {
                    error = `\`${part}\` variable name is not valid.`;
                    this.log.error(error);
                    valid = false;
                }
            }
        }

        this._validated = true;

        return { valid, error };
    }

    public render(variables?: ContextVariables): string {
        throw new Error('Code blocks rendering requires IFunctionRegistryReader. Incorrect method call.');
    }

    public async renderCode(context: SKContext): Promise<string> {
        if (!this._validated) {
            const { valid, error } = this.isValid();
            if (!valid) {
                throw new Error(error);
            }
        }

        this.log.trace(`Rendering code: "${this.content}"`);

        const parts = this.content.split(/[ \t\r\n]+/).filter((x) => x.trim() !== '');

        const functionName = parts[0];
        if (!context.registry) {
            throw new Error('Registry not set');
        }

        const fn = this.getFunctionFromRegistry(context.registry, functionName);
        if (!fn) {
            this.log.warn(`Function not found "${functionName}"`);
            return '';
        }

        // Using $input by default, e.g. when the syntax is {{functionName}}
        let funcParam = context.variables.input ?? '';
        if (parts.length > 1) {
            this.log.trace(`Passing required variable: "${parts[1]}"`);
            // If the code syntax is {{functionName $varName}} use $varName instead of $input
            funcParam = new VarBlock(parts[1], this.log).render(context.variables);
        }

        const result = await SKFunctionExtensions.invokeWithCustomInput(fn, new ContextVariables(funcParam), this.log);

        if (result.errorOccurred) {
            this.log.error(
                `Semantic function references a function "${functionName}" of incompatible type: defaulting to an empty result`
            );
            return '';
        }

        return result.result;
    }

    private getFunctionFromRegistry(registry: IFunctionRegistryReader, functionName: string): ISKFunction | undefined {
        // Search in the global space (only native functions there)
        if (registry.hasNativeFunction(functionName)) {
            return registry.getNativeFunction(functionName);
        }

        // If the function contains a skill name...
        if (functionName.includes('.')) {
            const functionNameParts = functionName.split('.');
            if (functionNameParts.length > 2) {
                this.log.error(`Invalid function name "${functionName}"`);
                throw new Error(
                    `Invalid function name "${functionName}". ` +
                        `A Function name can contain only one "." to separate skill name from function name.`
                );
            }

            const skillName = functionNameParts[0];
            functionName = functionNameParts[1];

            if (registry.hasNativeFunction(skillName, functionName)) {
                return registry.getNativeFunction(skillName, functionName);
            }

            if (registry.hasSemanticFunction(skillName, functionName)) {
                return registry.getSemanticFunction(skillName, functionName);
            }
        }

        return undefined;
    }
}

File: dotnet/src/Extensions/TemplateEngine/Blocks/CodeBlock.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.SemanticKernel.Diagnostics;
using Microsoft.SemanticKernel.Orchestration;
using Microsoft.SemanticKernel.SkillDefinition;

namespace Microsoft.SemanticKernel.TemplateEngine.Blocks;

#pragma warning disable CA2254 // error strings are used also internally, not just for logging
#pragma warning disable CA1031 // IsCriticalException is an internal utility and should not be used by extensions

// ReSharper disable TemplateIsNotCompileTimeConstantProblem
internal sealed class CodeBlock : Block, ICodeRendering
{
    internal override BlockTypes Type => BlockTypes.Code;

    /// <summary>
    /// Initializes a new instance of the <see cref="CodeBlock"/> class.
    /// </summary>
    /// <param name="content">Block content</param>
    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
    public CodeBlock(string? content, ILoggerFactory? loggerFactory)
        : this(new CodeTokenizer(loggerFactory).Tokenize(content), content?.Trim(), loggerFactory)
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="CodeBlock"/> class.
    /// </summary>
    /// <param name="tokens">A list of blocks</param>
    /// <param name="content">Block content</param>
    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
    public CodeBlock(List<Block> tokens, string? content, ILoggerFactory? loggerFactory)
        : base(content?.Trim(), loggerFactory)
    {
        this._tokens = tokens;
    }

    /// <inheritdoc/>
    public override bool IsValid(out string errorMsg)
    {
        errorMsg = "";

        foreach (Block token in this._tokens)
        {
            if (!token.IsValid(out errorMsg))
            {
                this.Logger.LogError(errorMsg);
                return false;
            }
        }

        if (this._tokens.Count > 1)
        {
            if (this._tokens[0].Type != BlockTypes.FunctionId)
            {
                errorMsg = $"Unexpected second token found: {this._tokens[1].Content}";
                this.Logger.LogError(errorMsg);
                return false;
            }

            if (this._tokens[1].Type is not BlockTypes.Value and not BlockTypes.Variable)
            {
                errorMsg = "Functions support only one parameter";
                this.Logger.LogError(errorMsg);
                return false;
            }
        }

        if (this._tokens.Count > 2)
        {
            errorMsg = $"Unexpected second token found: {this._tokens[1].Content}";
            this.Logger.LogError(errorMsg);
            return false;
        }

        this._validated = true;

        return true;
    }

    /// <inheritdoc/>
    public async Task<string> RenderCodeAsync(SKContext context, CancellationToken cancellationToken = default)
    {
        if (!this._validated && !this.IsValid(out var error))
        {
            throw new SKException(error);
        }

        this.Logger.LogTrace("Rendering code: `{0}`", this.Content);

        switch (this._tokens[0].Type)
        {
            case BlockTypes.Value:
            case BlockTypes.Variable:
                return ((ITextRendering)this._tokens[0]).Render(context.Variables);

            case BlockTypes.FunctionId:
                return await this.RenderFunctionCallAsync((FunctionIdBlock)this._tokens[0], context).ConfigureAwait(false);
        }

        throw new SKException($"Unexpected first token type: {this._tokens[0].Type:G}");
    }

    #region private ================================================================================

    private bool _validated;
    private readonly List<Block> _tokens;

    private async Task<string> RenderFunctionCallAsync(FunctionIdBlock fBlock, SKContext context)
    {
        if (context.Skills == null)
        {
            throw new SKException("Skill collection not found in the context");
        }

        if (!this.GetFunctionFromSkillCollection(context.Skills!, fBlock, out ISKFunction? function))
        {
            var errorMsg = $"Function `{fBlock.Content}` not found";
            this.Logger.LogError(errorMsg);
            throw new SKException(errorMsg);
        }

        SKContext contextClone = context.Clone();

        // If the code syntax is {{functionName $varName}} use $varName instead of $input
        // If the code syntax is {{functionName 'value'}} use "value" instead of $input
        if (this._tokens.Count > 1)
        {
            // Sensitive data, logging as trace, disabled by default
            this.Logger.LogTrace("Passing variable/value: `{0}`", this._tokens[1].Content);

            string input = ((ITextRendering)this._tokens[1]).Render(contextClone.Variables);
            // Keep previous trust information when updating the input
            contextClone.Variables.Update(input);
        }

        Exception? localException = null;
        try
        {
            contextClone = await function!.InvokeAsync(contextClone).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            this.Logger.LogError(ex, "Something went wrong when invoking function with custom input: {0}.{1}. Error: {2}",
                function!.SkillName, function.Name, ex.Message);
            localException = ex;
        }

        if (contextClone.ErrorOccurred || localException is not null)
        {
            var lastException = localException ?? contextClone.LastException;
            var errorMsg = $"Function `{fBlock.Content}` execution failed. {lastException?.GetType().FullName}: {lastException?.Message}";
            this.Logger.LogError(errorMsg);
            throw new SKException(errorMsg, lastException);
        }

        return contextClone.Result;
    }

    private bool GetFunctionFromSkillCollection(
        IReadOnlySkillCollection skills,
        FunctionIdBlock fBlock,
        out ISKFunction? function)
    {
        if (string.IsNullOrEmpty(fBlock.SkillName))
        {
            // Function in the global skill
            return skills.TryGetFunction(fBlock.FunctionName, out function);
        }

        // Function within a specific skill
        return skills.TryGetFunction(fBlock.SkillName, fBlock.FunctionName, out function);
    }

    #endregion
}
// ReSharper restore TemplateIsNotCompileTimeConstantProblem
#pragma warning restore CA2254
